prov/efa/src/efa_mr.c

```diff
@@ -511,76 +521,67 @@ struct ibv_mr *efa_mr_reg_ibv_dmabuf_mr(struct ibv_pd *pd, uint64_t offset,
 					int access, const uint64_t flags)
 {
-	if (flags & FI_MR_DMABUF)
-		return efa_mr_reg_ibv_dmabuf_mr(
-			efa_mr->domain->ibv_pd,
-			mr_attr->dmabuf->offset,
-			mr_attr->dmabuf->len,
-			(uintptr_t) mr_attr->dmabuf->base_addr + mr_attr->dmabuf->offset,
-			mr_attr->dmabuf->fd,
-			access
-		);
+	int ret, dmabuf_fd;
+	struct ibv_mr *mr = NULL;
+	uint64_t iova, offset;
+	size_t len;
+	enum efa_dmabuf_support_status *dmabuf_support;
+
+	dmabuf_support = &(efa_mr->domain->hmem_info[mr_attr->iface].dmabuf_support_status);
+
+	if (flags & FI_MR_DMABUF) {
+		goto dmabuf;
+	}
+
+	if (*dmabuf_support == EFA_DMABUF_NOT_SUPPORTED) {
+		goto no_dmabuf;
+	}
+dmabuf:
+	if (flags & FI_MR_DMABUF) {
+		dmabuf_fd = mr_attr->dmabuf->fd;
+		offset = mr_attr->dmabuf->offset;
+		iova = (uintptr_t) mr_attr->dmabuf->base_addr +
+		       mr_attr->dmabuf->offset;
+		len = mr_attr->dmabuf->len;
+	} else {
+		iova = (uint64_t) mr_attr->mr_iov->iov_base,
+		len = mr_attr->mr_iov->iov_len;
+		ret = ofi_hmem_get_dmabuf_fd(mr_attr->iface, (void *)iova, len,
+		                             &dmabuf_fd, &offset);
 
-	/*
-	 * TODO: remove the synapseai and neuron blocks by onboarding the
-	 * ofi_hmem_get_dmabuf_fd API.
-	 */
-#if HAVE_SYNAPSEAI
-	if (efa_mr_is_synapseai(efa_mr)) {
-		int dmabuf_fd;
-		uint64_t offset;
-		int ret;
-
-		ret = synapseai_get_dmabuf_fd(mr_attr->mr_iov->iov_base,
-						(uint64_t) mr_attr->mr_iov->iov_len,
-						&dmabuf_fd, &offset);
 		if (ret != FI_SUCCESS) {
-			EFA_WARN(FI_LOG_MR, "Unable to get dmabuf fd for Gaudi device buffer \n");
-			return NULL;
+			EFA_WARN(FI_LOG_MR,
+				 "Unable to get dmabuf fd for device buffer. "
+				 "errno: %d, err_msg: %s\n",
+				 ret, fi_strerror(-ret));
+			goto fallback;
 		}
-		return efa_mr_reg_ibv_dmabuf_mr(efa_mr->domain->ibv_pd, offset,
-					mr_attr->mr_iov->iov_len,
-					(uint64_t)mr_attr->mr_iov->iov_base,
-					dmabuf_fd, access);
 	}
-#endif
 
-#if HAVE_NEURON
-	if (efa_mr_is_neuron(efa_mr)) {
-		int dmabuf_fd;
-		uint64_t offset;
-		int ret;
-
-		ret = neuron_get_dmabuf_fd(
-				mr_attr->mr_iov->iov_base,
-				mr_attr->mr_iov->iov_len,
-				&dmabuf_fd,
-				&offset);
-
-		if (ret == FI_SUCCESS) {
-			/* Success => invoke ibv_reg_dmabuf_mr */
-			return efa_mr_reg_ibv_dmabuf_mr(
-					efa_mr->domain->ibv_pd, 0,
-					mr_attr->mr_iov->iov_len,
-					(uint64_t)mr_attr->mr_iov->iov_base,
-					dmabuf_fd, access);
-		} else if (ret == -FI_ENOPROTOOPT) {
-			/* Protocol not availabe => fallback */
-			EFA_INFO(FI_LOG_MR,
-				"Unable to get dmabuf fd for Neuron device buffer, "
-				"Fall back to ibv_reg_mr\n");
-			return ibv_reg_mr(
-				efa_mr->domain->ibv_pd,
-				(void *)mr_attr->mr_iov->iov_base,
-				mr_attr->mr_iov->iov_len, access);
-		}
-		return NULL;
+	EFA_INFO(FI_LOG_MR,
+		 "Registering dmabuf mr with fd: %d, offset: %lu, len: %zu\n",
+		 dmabuf_fd, offset, len);
+
+	mr = efa_mr_reg_ibv_dmabuf_mr(efa_mr->domain->ibv_pd, offset, len, iova,
+				      dmabuf_fd, access);
+
+fallback:
```

Comment:
We need some more discussion here. I am not sure this fallback logic is the best practice.

Reply:
Yep let's have a discussion.

Reply:
For CUDA we are not sure about the correct fallback behavior if dmabuf fd/mr fails. Therefore I removed the fallback logic for CUDA in update. The application must request FI_MR_DMABUF with a valid fd in order to use dmabuf MR.